doc/Running and Writing Tests for Sia.md

```diff
@@ -0,0 +1,278 @@
+of the code is up to date. New code should be on its own branch (again, see
+[doc/Guide to Contributing to Sia.md][guide]).
+
+```bash
+# Make sure you are in the right directory.
+$ cd $GOPATH/src/github.com/<your Github username>/Sia
+# Checkout the branch you made the changes on.
+$ git checkout <branch name>
+# Stash any tracked but uncommitted changes.
+$ git stash
+# Then switch back to `master` and update it to match the original repo.
+$ git checkout master
+$ git pull upstream master
+# Update `origin`, your fork of the repo.
+$ git push origin master
+# Make the updated `master` the new base of the branch you made the changes on,
+# which involves reapplying all the commits made to that branch.  Without the
+# `--ignore-date` flag, git rebase changes the date on all the commits to the
+# current date.
+$ git checkout <branch name>
+$ git rebase master --ignore-date
+# Restore the changes you stashed earlier.
+$ git stash pop
+```
+When you call `rebase`, you may run into some merge conflicts.  Luke Champine's
+['How to into git and GitHub'][luke] has more details (and many useful tricks).
+
+Once the branch you want to test is up to date, you're ready to run some tests.
+
+<a name="entire"/>
+### Testing the entire build
+The `make test` command runs all tests (functions starting with `Test` in
+`_test.go` files) for each package, setting off a panic for any test that runs
+longer than 5s.  For verbose output, run `make test-v` (which panics after 15s
+instead of 5s).  Finally, `make test-long` has verbose output, only panics when
+a test takes 5 minutes, and also cleans up your code using `gofmt` and `golint`.  
+**You should run** `make test-long` **before each pull request.**
+
+Run `make cover` to run all tests for each package and generate color-coded
+.html visualizations of test coverage by function for each source file.  Open
+`cover/<module>.html` in a browser to inspect a module's test coverage.
+
+Meanwhile, `make bench` will call `gofmt` on all packages, then run all
+benchmarks (functions starting with `Benchmark` in `_test.go` files).
+
+<a name="particular"/>
+### Testing a particular package
+It can take a while to run the custom `make` commands defined in Sia's Makefile,
+since they check all packages.  If you just want to check the package you've
+made changes in, you can run the commands below.  But don't forget to run `make
+test-long` before making a pull request!
+
+```bash
+# Make sure you are in the right directory.
+$ cd $GOPATH/src/github.com/<your Github username>/Sia
+# Run all tests in directory <package> and view test coverage for <package>.
+$ go test -tags='testing debug' -timeout=360s -covermode=atomic -coverprofile=cover/<package>.out ./<package>
+$ go tool cover -html=cover/<package>.out -o=cover/<package>.html
+$ rm cover/<package>.out
+```
+Now you can open `cover/<package>.html` in a browser to view test coverage for
+<package>.  On OS X, you can just do `open cover/<package.html>`.
```

Comment:
Actually, it's much simpler than that. To test a single package, run `make test pkgs=./path/to/package`. Ditto for `make bench` and `make cover`. Furthermore, you can test a single function using `make test pkgs=./package run=TestFoo`. This is made possible by the `pkgs` and `run` variables in the Makefile.